### ISSUE TITLE
vm_compute does not support universe-polymorphic definitions

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#4247]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 4247
Date: 2015-05-29 15:29:36 +0200
From: @_gares
To:   
CC:   Benjamin.Gregoire@sophia.inria.fr, coq-bugs-redist@lists.gforge.inria.fr, @_gmalecha, @_gmalecha, @_silene, @_JasonGross, @_mattam82

Last updated: 2015-10-28 17:06:07 +0100

## Bugzilla Comment ID: 10414
Date: 2015-05-29 15:29:36 +0200
From: @_gares

The test success/univscompute.v fails if I configure with -no-native-compiler while works if I don't use such flag.

No idea if the bug pertains to vm_compute or native_compiler... or to poly univs.
It is not an April's fool.  I've tried it twice, I was not believing my eyes ;-)

Here the quickest way to reproduce it:

./configure -local -no-native-compiler && make -j2 states && rm -f test-suite/success/univscompute.v.log && make -C test-suite/ success/univscompute.v.log && cat test-suite/success/univscompute.v.log

Is my PC cursed?

## Bugzilla Comment ID: 10415
Date: 2015-05-29 15:36:11 +0200
From: @_gares

OK, I'm silly, the test calls native_compute but the error message is far from being "disabled at compilation time".  It is an error about missing universe constraints I guess.  I hope native_compute falls back to vm_compute if -no-native-compute is passed (so that my bug report is still valid ;-)

## Bugzilla Comment ID: 10448
Date: 2015-06-09 15:11:15 +0200
From: @_maximedenes

When using -no-native-compiler, calls to native_compute indeed fallback to vm_compute. Replacing native_compute by vm_compute in the file also triggers the problem, which is orthogonal to native_compute.

## Bugzilla Comment ID: 10449
Date: 2015-06-09 17:01:06 +0200
From: @_gares

Right, so the component field "VM" is right.
I suspect it is related to the patch by Matthieu w.r.t. poly univs.

I guess the best option is to say that Matthieu and Benjamin will chat about that here in 2 weeks.

I'm assigning it to Matthieu.

Best,

## Bugzilla Comment ID: 10664
Date: 2015-07-20 15:38:35 +0200
From: @_silene

*** Bug #4263 has been marked as a duplicate of this bug. ***

## Bugzilla Comment ID: 10665
Date: 2015-07-20 15:44:36 +0200
From: @_silene

As reported in bug #4263, one of the issues is that vm_compute does not unfold universe-polymorphic definitions:

Polymorphic Definition id {T : Type} (x : T) := x.
Eval vm_compute in @id Prop.

     = id
     : Prop -> Prop

## Bugzilla Comment ID: 10667
Date: 2015-07-20 19:00:40 +0200
From: @_mattam82

I couldn't get time to let benjamin help us with this, this has to be postponed to 8.6. The missing feature is already documented in the manual.

## Bugzilla Comment ID: 10672
Date: 2015-07-21 09:42:46 +0200
From: @_gmalecha

I was emboldened by my experience at the coding sprint to try to get something working here. It was the project that I proposed after all and the lack of functionality is making my computational reflection code useless in trunk.

Let me start by saying that this patch is pretty ugly (due in part to knowing relatively little about the VM 14 hours ago); however, it does work on the examples that I've run it through (i.e. the polymorphic identity function and a few variants). The idea is to introduce an extra lambda at the beginning of polymorphic functions that will carry the universes, and a corresponding application when these terms are applied. Using this strategy, the only problem comes from parsing terms at the end especially when we have polymorphic constants that are opaque since there is not a convenient way to rebuild them appropriately. In the current strategy, this would need a special case inside the code generator.

Anyways, the patch it attached (it is based on the v8.5 branch, but should be relocatable to trunk). I'm happy to dig more into the implementation if a variation of this strategy is even remotely acceptable. Otherwise, I'm happy to dig into alternative suggestions.

## Bugzilla Comment ID: 10673
Date: 2015-07-21 09:43:32 +0200
From: @_gmalecha

Created attachment 608
patch (on v8.5)

> Attached file: vm-poly.patch (text/plain, 16738 bytes)
> Description:   patch (on v8.5)

## Bugzilla Comment ID: 10690
Date: 2015-07-22 09:59:27 +0200
From: @_silene

I have no idea if this is the proper way of doing it (or if it is even correct), so I will not comment on it. (I guess you could look how native_compute achieves it. If it does it the same way, that would be a good clue to the correctness of your approach.)

Nevertheless, if you submit a separate patch with only your comments (and possibly cleanups) to the bytecode engine, I am willing to apply it. Just make sure that the comments to cbytecodes.ml match the ones from cbytecodes.mli (or just remove the comments from cbytecodes.ml).

## Bugzilla Comment ID: 10694
Date: 2015-07-22 16:00:59 +0200
From: @_mattam82

I have no idea if this is correct either, but the idea sounds fine and I'd be happy to try and integrate it in 8.5. I think we'll have time to have a look at it tomorrow with Maxime who knows the VM better than I do. Dealing with opaque constants might or might not be difficult. Basically you should come to a situation where your value is a constant and the stack contains the universe instance as Type@{i's} no ? Is the problem finding out which subset of the stack to use ? It is something else ? I think the encoding with Lambda and applications can be made to work, but I don't see how you deal with the Var(k)'s that make up the instances in the bodies of polymorphic constants which are de Bruijn levels. Could you explain compile_getglobal/_universe_instance in more details?

## Bugzilla Comment ID: 10695
Date: 2015-07-22 16:02:12 +0200
From: @_mattam82

I also agree a separate patch with your comments (there's an acc -> accu typo there) would be nice.

## Bugzilla Comment ID: 10701
Date: 2015-07-22 21:28:16 +0200
From: @_gmalecha

Related to the separate commit. Is it prefereable to just document the bytecodes in the .mli or should I copy the comments to the .ml?

@Matthieu:
Basically you should come to a situation where your value is a constant and the stack contains the universe instance as Type@{i's} no ? Is the problem finding out which subset of the stack to use ? It is something else ?

I didn't dig into this code too much, but I agree that it shouldn't be too difficult. The annoying bit is that the current virtual machine uses [pconstant] for the representation of constants. What it should use is [constant] so that the pieces can be assembled after the fact (since the universes can contain variables, your next question/point). Since the "parsing" phase is type-directed, we need a check that says that if the constant is polymorphic then the first argument should be interpreted as the instance arguments to construct a [pconstant] from a [pconstant] (for historical reasons?) and an [instance]. 

I think the encoding with Lambda and applications can be made to work, but I don't see how you deal with the Var(k)'s that make up the instances in the bodies of polymorphic constants which are de Bruijn levels. Could you explain compile_getglobal/_universe_instance in more details?

On further inspection of the code universe polymorphism implementation, I think that my assumption was wrong. But I have an idea of how to fix it, and I'll plan on doing that shortly.

## Bugzilla Comment ID: 10703
Date: 2015-07-23 07:02:02 +0200
From: @_gmalecha

So I kept looking into this. I have fixed some of the code but it seems like the thing that is missing is a way to represent universes in the byte-code. For that, I'm not certain which representation of universes we will have. There is already a way to store [Term.sort] within the byte-code and hooking into that seems like it would require a minimal amount of extra work but that doesn't seem like exactly what goes into the [universe_instance] structure that we need to make in order to turn a [constant] into a [pconstant].

The most current code is here: https://github.com/gmalecha/coq/tree/univ-compute

I submitted a pull-request that has just the comments on the vm.

## Bugzilla Comment ID: 10704
Date: 2015-07-23 07:45:34 +0200
From: @_silene

This will show my poor understanding of the issue, but there is something I don't understand with polymorphic definitions. (I guess we will discuss about it face to face, but here it is anyway.) Why do universe levels matter at reduction time? It seems to me that the whole work has been done at (pre)typing time. All the polymorphic constants should already be instantiated with proper levels by the time we get to vm_compute, so why do polymorphic constants need some special handling? It is not like the reduction could create some fresh universe variables, is it?

## Bugzilla Comment ID: 10705
Date: 2015-07-23 08:17:05 +0200
From: @_gmalecha

The problem lies with polymorphic instantiation. Here's an example:

Polymorphic Axiom foo : forall T:Type, T -> T.

Polymorphic Definition idp (T : Type) (x : T) : T := foo x.

The result of

Eval vm_compute in idp nat 3.

should be

foo@{Set} nat 3.

we need to track universes through the computation so that we know that that should be foo@{Set} and not foo@{Type} or some other universe. The current vm avoids this by never reducing in any context that has quantified universes so it doesn't need to do substitution.

## Bugzilla Comment ID: 10706
Date: 2015-07-23 10:20:07 +0200
From: @_silene

Your example is too simple to convince me and I cannot envision a better example yet.

If you were to do a reduction without caring about universe levels, you would still get "foo nat 3". If you need to recover the universe levels hidden in foo, you could just look at the arguments of foo (a simple instantiation of foo, outside the vm). But do we even need to recover these universe levels? It is certainly not for consistency reasons, since all the universe constraints have already been generated at the time "idp nat 3" was typed. So what are the reasons and do they require specific vm support?

Perhaps another way to view it: universe levels have absolutely no computational content, so why should the vm ever deal with them?

## Bugzilla Comment ID: 10708
Date: 2015-07-24 08:38:14 +0200
From: @_silene

We had a small meeting on this point yesterday. First, forget what I said; there are indeed some cases where the universe levels computed by vm_compute actually matters. Second, here are two possible implementations we discussed. Let us suppose you want to compile to byte-code:

Polymorphic Definition foo@{i j k} (x y z : ...) := ... (bar@{i k} y) ...

Solution 1: have one binder per universe and handle them exactly as if they were traditional non-universe function arguments. You would compile foo as if it was

fun i j k x y z -> ... (bar i k y) ...

Note that the universe part would always be fully applied. At the end, any opaque symbol baz@{u v} would have as its two first arguments all the universe levels, and as the following arguments the standard arguments (some might be missing, if partially applied).

Type@{i} itself would compiled to a dummy polymorphic opaque constant equal to Type. When decompiling at the end, one just has to recognize this opaque constant and gets its first argument to obtain the universe level.

Solution 2: have one binder to an array of universe levels, but compute this array once and for all in a closure. You would compile foo as if it was

fun u ->
  let bar_17 = bar [|u.(0); u.(2)|] in
  ...
  fun x y z -> ... (bar_17 y) ...

That way, the execution cost of polymorphic definitions is almost the same as for monomorphic ones, at the expense of a startup overhead of vm_compute when all the needed closures are created.

We did not discuss it yesterday, but it seems to me that solution 2 means that all the definitions have to be compiled using the above scheme to ensure that closures are created only once per execution. In particular, it means that monomorphic definitions have to start by "fun () ->". The only definitions that could use the old compilation scheme are the ones that do not call any polymorphic definitions, not even transitively!

Also, some extra care has to be taken when compiling mutually recursive definitions, so that the creation of closures actually terminates. But I guess it would be obvious when implementing it.

So, to summarize, solution 1 is much simpler while solution 2 would hopefully leads to better performances. An intermediate solution where an array of levels is built dynamically at each call is thought to be the worst of both worlds, due to the sheer amount of allocations it would cause.

## Bugzilla Comment ID: 10808
Date: 2015-08-13 23:11:11 +0200
From: @_gmalecha

(Just in case anyone else is thinking about the problem concurrently)

I have been making slow progress through the implementation, but it isn't quite there yet. The code is currently available here:

https://github.com/gmalecha/coq/tree/v8.5-vm-univ

## Bugzilla Comment ID: 10855
Date: 2015-09-04 02:25:55 +0200
From: @_gmalecha

My implementation has made progress. My code can handle compiling of universe polymorphic functions and axioms and correctly propagates universes through computation.

I've been able to handle all of this without adding anything too fancy to the vm compiler (the VM itself is unchanged). What I am working on now is instantiating types. For example:

Polymorphic Definition xxx := Type.

Eval vm_compute in xxx.

Here, I have to be able to instantiate Type which is not represented in the same way as a constant. I am wondering if anyone can point me to the right way to do this. Ideally it seems like I could directly construct a Universe.t in the VM, but I'm not exactly certain how to make that work.

Code is still available at https://github.com/gmalecha/coq/tree/v8.5-vm-univ recently rebased on top of trunk.

## Bugzilla Comment ID: 10856
Date: 2015-09-04 07:03:34 +0200
From: @_silene

There was a proposal for this specific problem in comment 17: do not directly use the Type sort constructor, but instead assume that there exists an opaque polymorphic constant. That way, Type can be represented like any other constant. In other words,

Polymorphic Definition xxx := Type.

could be compiled into

(* Polymorphic Axiom __Type : Type. *)
Polymorphic Definition xxx := __Type.

That way, you do not have to modify the VM itself. Compilation has to be trivially modified to replace any occurrence of the sort "Type" by the constant "__Type". Decompilation performs the opposite transformation; since __Type is a polymorphic accumulator, the universe levels are immediately available as its arguments.

## Bugzilla Comment ID: 10857
Date: 2015-09-04 07:06:48 +0200
From: @_gmalecha

I assume then that that constant should go somewhere in the prelude? There is no way to make it visible only by the byte-code is there?

## Bugzilla Comment ID: 10858
Date: 2015-09-04 07:17:07 +0200
From: @_silene

It would be better if the constant was not part of the prelude (that is, if it was hardcoded in the bytecode compiler). I do not know how hard that would be to implement though, due to potential serialization issues.

## Bugzilla Comment ID: 10859
Date: 2015-09-04 15:24:11 +0200
From: @_mattam82

You could maybe add it to the environment used to do vm_compute if it's used functionally? Adding a Universe.t to the VM would be cleaner I suppose, couldn't you treat sorts like axioms there? They would just come with a universe instance that can be substituted into.

## Bugzilla Comment ID: 10884
Date: 2015-09-09 22:32:59 +0200
From: @_gmalecha

What is the right way to substitute variables inside of a Univ.universe? When I am compiling constrs I have to handle Type@{e} where 'e' is a Univ.universe. If 'e' contains variables then I need to substitute them when I instantiate. An example would be something like:

forall x : Type@{Ui+1}, ...

Or are terms like this not allowed? In that case, what prevents them from being constructed?

## Bugzilla Comment ID: 10885
Date: 2015-09-09 22:48:47 +0200
From: @_JasonGross

Empirically, Type@{i+1} is forbidden by the parser, but Type@{max(i,j)} is perfectly valid:

Coq < Set Printing Universes.

Coq < Check Type@{max(i,j)}.
Type@{max(i, j)}
     : Type@{max(i+1, j+1)}
(* Top.3
   Top.4 |=  *)

Coq < Check Type@{i+1}.
Toplevel input, characters 13-14:
> Check Type@{i+1}.
>              ^
Syntax error: '}' expected after [universe] (in [constr:sort]).

## Bugzilla Comment ID: 10886
Date: 2015-09-10 03:19:13 +0200
From: @_gmalecha

Is there any place where I can find test cases that make extensive use of both universe polymorphism and computation? I need my patch to go through some more rigorous testing before I clean it up.

## Bugzilla Comment ID: 10887
Date: 2015-09-10 04:49:27 +0200
From: @_JasonGross

The stuff involving duality in the category theory part of the HoTT library stresses, to some degree, computation to unify types, universes, and primitive projections.  In particular, https://github.com/HoTT/HoTT/blob/master/theories/categories/Comma/Dual.v used to take around 5s to typecheck two definitions (no tactics); now it takes (http://hott.github.io/HoTT/timing-html/HoTT.categories.Comma.Dual.html) 0.1s on the first definition and 0.6s on the second definition, which is basically just [Definition foo : bar := (eq_refl, eq_refl).]  (I know these aren't slow by your standards...)

## Bugzilla Comment ID: 10888
Date: 2015-09-10 08:04:55 +0200
From: @_gmalecha

I'll be interested in performance at some point (since there are two proposals for compiling universes), but at the moment I'm still focused on correctness. Is this part of HoTT run as part of the commit tests?

@Matthieu, is there a place that I can get your universe polymorphic version of the standard library to test on? Something like that seems ideal since it probably has relatively few dependencies which makes iteration substantially faster.

## Bugzilla Comment ID: 10889
Date: 2015-09-10 08:23:16 +0200
From: @_JasonGross

I was mostly using timing as a measure of how much it stressed the computation machinery.

> Is this part of HoTT run as part of the commit tests?

I'm not sure what you mean by "commit tests", but it's part of the `make all` (or just `make`) target, and it's automatically run on Travis CI when new commits are added.  (You can also just run `./autogen.sh && ./configure && make theories/categories/Comma/Dual.vo`.)

If you want another part that stresses universes (but not so much computation, if I recall correctly), there's hit/V.v, which, last I checked, used between 20,000 and 40,000 universe variables.

## Bugzilla Comment ID: 10931
Date: 2015-09-16 19:35:58 +0200
From: @_gmalecha

I have opened pull-request #103 with the current code. Seeking feedback on the patch at this point.

Thanks.

## Bugzilla Comment ID: 10933
Date: 2015-09-17 13:07:48 +0200
From: @_mattam82

Hi Gregory, I'll have a look at it in the coming week, it was agreed at the last WG to put your patch in the next beta.

## Bugzilla Comment ID: 11141
Date: 2015-10-28 17:06:07 +0100
From: @_maximedenes

Fixed in v8.5. Thanks a lot for your contribution, Greg!
